clone/cmd/sumdbclone/internal/client/sumdb.go

```diff
@@ -0,0 +1,164 @@
+const HashLenBytes = 32
+
+// pathBase is the max number of entries in any SumDB directory.
+// Beyond this, sub directories are created.
+// https://github.com/golang/mod/blob/346a37af5599be02f125bd8bc0a5e1d33db21ddc/sumdb/tlog/tile.go#L168
+const pathBase = 1000
+
+// Fetcher gets data paths. This allows impl to be swapped for tests.
+type Fetcher interface {
+	// GetData gets the data at the given path, or returns an error.
+	GetData(path string) ([]byte, error)
+}
+
+// Checkpoint is a parsed Checkpoint and the raw bytes that it came from.
+type Checkpoint struct {
+	*tlog.Tree
+
+	Raw []byte
+}
+
+// SumDBClient provides access to information from the Sum DB.
+type SumDBClient struct {
+	height  int
+	vkey    string
+	fetcher Fetcher
+}
+
+// NewSumDB creates a new client that fetches tiles of the given height.
+func NewSumDB(height int, vkey, url string, c *http.Client) *SumDBClient {
+	return &SumDBClient{
+		height: height,
+		vkey:   vkey,
+		fetcher: &HTTPFetcher{
+			c:       c,
+			baseURL: url,
+		},
+	}
+}
+
+// LatestCheckpoint gets the freshest Checkpoint.
+func (c *SumDBClient) LatestCheckpoint() (*Checkpoint, error) {
+	checkpoint, err := c.fetcher.GetData("/latest")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get /latest Checkpoint; %w", err)
+	}
+
+	verifier, err := note.NewVerifier(c.vkey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create verifier: %w", err)
+	}
+	verifiers := note.VerifierList(verifier)
+
+	note, err := note.Open(checkpoint, verifiers)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open note: %w", err)
+	}
+	tree, err := tlog.ParseTree([]byte(note.Text))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse tree: %w", err)
+	}
+
+	return &Checkpoint{Tree: &tree, Raw: checkpoint}, nil
+}
+
+// FullLeavesAtOffset gets the Nth chunk of 2**height leaves.
+func (c *SumDBClient) FullLeavesAtOffset(offset int) ([][]byte, error) {
+	data, err := c.fetcher.GetData(fmt.Sprintf("/tile/%d/data/%s", c.height, c.tilePath(offset)))
+	if err != nil {
+		return nil, err
+	}
+	return dataToLeaves(data), nil
+}
+
+// PartialLeavesAtOffset gets the final tile of incomplete leaves.
+func (c *SumDBClient) PartialLeavesAtOffset(offset, count int) ([][]byte, error) {
+	data, err := c.fetcher.GetData(fmt.Sprintf("/tile/%d/data/%s.p/%d", c.height, c.tilePath(offset), count))
+	if err != nil {
+		return nil, err
+	}
+	return dataToLeaves(data), nil
+}
+
+// tilePath constructs the component of the path which refers to a tile at a
+// given offset. This was copied from the core implementation:
+// https://github.com/golang/mod/blob/346a37af5599be02f125bd8bc0a5e1d33db21ddc/sumdb/tlog/tile.go#L171
+func (c *SumDBClient) tilePath(offset int) string {
+	nStr := fmt.Sprintf("%03d", offset%pathBase)
+	for offset >= pathBase {
+		offset /= pathBase
+		nStr = fmt.Sprintf("x%03d/%s", offset%pathBase, nStr)
+	}
+	return nStr
+}
+
+func dataToLeaves(data []byte) [][]byte {
```

Comment:
Is this the same as `bytes.Split(data, []byte{'\n'})`?

Reply:
No, but I've replaced it with something similar. Annoying to split on "\n\n" and then need to re-add it, but I'll survive.